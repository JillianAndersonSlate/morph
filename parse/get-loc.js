export default (line, scolumn, ecolumn) => ({
  start: {
    line,
    column: scolumn,
  },
  end: {
    line,
    column: ecolumn,
  },
})
