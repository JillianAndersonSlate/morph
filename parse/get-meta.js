import { getCodeData, isCode } from './helpers.js'

export default (value, line, startLine) => {
  if (!isCode(value)) return []

  return getCodeData(value).map(value => ({
    tag: 'code',
    value,
  }))
}
