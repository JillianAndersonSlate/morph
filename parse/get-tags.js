import { isShorthand, isStyle, isProps } from './helpers.js'

const CODE_PROPS = ['from', 'when', 'onClick', 'onFocus', 'onWhen']
const shouldBeProps = prop => CODE_PROPS.includes(prop) || /^on[A-Z]/.test(prop)

export default (name, props, value, block) => {
  const tags = {}

  if (isStyle(name)) tags.style = true
  if (isShorthand(name) && block.isBasic) {
    tags.shorthand = true
  }

  if (shouldBeProps(name)) tags.shouldBeProps = true
  if (props) tags.props = true

  if (tags.props || tags.shouldBeProps) {
    tags.validProps = isProps(props)
  }

  return tags
}
