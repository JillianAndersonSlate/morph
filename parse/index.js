import {
  getBlock,
  getFontInfo,
  getProp,
  getSection,
  getTodo,
  getValue,
  isBasic,
  isBlock,
  isCapture,
  isEmptyList,
  isEnd,
  isFontable,
  isGroup,
  isItem,
  isList,
  isProp,
  isSection,
  isTodo,
  stemStylesFromProp,
  warn,
} from './helpers.js'
import getLoc from './get-loc.js'
import getMeta from './get-meta.js'
import getTags from './get-tags.js'

export default text => {
  const fonts = {}
  const lines = text.split('\n')
  const stack = []
  const todos = []
  const views = []
  let lastCapture

  const lookForFonts = block => {
    if (block.properties && (isFontable(block.name.value) || !block.isBasic)) {
      let fontFamily
      let fontWeight

      block.properties.list.forEach(p => {
        if (p.key.value === 'fontFamily') {
          fontFamily = p.value.value
        } else if (p.key.value === 'fontWeight') {
          fontWeight = p.value.value
        }
      })

      if (fontFamily) {
        const info = getFontInfo(fontFamily, fontWeight)

        info.forEach(font => {
          if (!fonts[font.family]) fonts[font.family] = []
          if (!fonts[font.family].includes(font.weight)) {
            fonts[font.family].push(font.weight)
          }
        })
      }
    }
  }

  const end = (block, maybeEndLine) => {
    let endLine = maybeEndLine
    while (lines[endLine] === '' && endLine > 0) {
      endLine--
    }

    const prevLine = lines[endLine - 1]

    block.loc.end = {
      line: endLine,
      column:
        (typeof prevLine === 'string' ? prevLine : lines[endLine]).length - 1,
    }
    // TODO review this
    // if we're past our line, use the previous line's end
    if (block.loc.end.column < 0) {
      block.loc.end = {
        line: endLine - 1,
        column: lines[endLine - 2].length - 1,
      }
    }

    // TODO define end location of blocks inside this
    // block.blocks.forEach(innerBlock => {
    //   innerBlock.loc.end = // ...
    // })
    // end this block's properties if any
    if (block.properties) {
      const last = block.properties.list[block.properties.list.length - 1]
      block.properties.loc.end = last.loc.end

      // look for fonts
      lookForFonts(block)
    }

    if (block.blocks) {
      block.blocks.list.forEach(lookForFonts)
    }

    if (stack.length > 0) {
      // // if there are more blocks, put this block as part of the last block on the stack's block
      // const nextLast = stack[stack.length - 1]
      // if (nextLast !== block && nextLast.blocks) {
      //   nextLast.blocks.list.push(block)
      // }
      return false
    } else {
      // if we're the last block on the stack, then this is the view!
      views.push(block)
      return true
    }
  }

  const parseBlock = (l, i, line, lineIndex) => {
    const { block: name, is } = getBlock(line)
    let shouldPushToStack = false

    const block = {
      type: 'Block',
      name: {
        type: 'Literal',
        value: name,
        loc: getLoc(lineIndex, l.indexOf(line), l.length - 1),
      },
      isBasic: isBasic(name),
      loc: getLoc(lineIndex, l.indexOf(line)),
      parents: stack
        .filter(b => b.type === 'Block')
        .map(b => b.is || b.name.value),
    }
    if (is) {
      block.is = is

      if (isCapture(name)) {
        if (lastCapture) {
          lastCapture.captureNext = is
        }
        lastCapture = block
      }
    }

    const last = stack[stack.length - 1]
    if (last) {
      if (last.blocks) {
        if (isList(last.name.value) && last.blocks.list.length > 0) {
          // the list already has a block
          let topLevelIsGroup = false
          if (stack.length - 2 >= 0) {
            topLevelIsGroup = isGroup(stack[stack.length - 2].name.value)
          } else if (views.length > 0) {
            topLevelIsGroup = !!views[0].blocks
          }

          if (topLevelIsGroup) {
            // tell how we can fix that
            warn(
              lines[i - 1] === ''
                ? `put 1 empty line before`
                : `put 2 empty lines before`,
              block
            )
          } else {
            warn(`add Vertical at the top`, block)
          }
          // shouldPushToStack = true
        }

        last.blocks.list.push(block)
      } else {
        // the block is inside a block that isn't a group
        end(stack.pop(), lineIndex - 1)
        const topLevelIsGroup = !!views[0].blocks

        if (topLevelIsGroup) {
          // tell how we can fix that
          warn(
            lines[i - 1] === ''
              ? `put 1 empty line before`
              : `put 2 empty lines before`,
            block
          )
        } else {
          warn(`add Vertical at the top`, block)
        }
        shouldPushToStack = true
      }
    } else if (views.length > 0) {
      // the block is outside the top level block
      const topLevelIsGroup = !!views[0].blocks
      let newLinesBeforePreviousBlock = 1
      while (isEnd(lines[i - newLinesBeforePreviousBlock])) {
        newLinesBeforePreviousBlock++
      }

      const help = []
      if (!topLevelIsGroup) {
        help.push(`add Vertical at the top`)
      }
      if (newLinesBeforePreviousBlock > 2) {
        const linesToRemove = newLinesBeforePreviousBlock - 2
        help.push(
          `remove ${linesToRemove} empty line${linesToRemove > 1
            ? 's'
            : ''} before`
        )
      }
      warn(help.join(', '), block)
    }

    if (isGroup(name)) {
      block.blocks = {
        type: 'Blocks',
        list: [],
        loc: getLoc(lineIndex + 1, 0),
      }
      shouldPushToStack = true
    }

    if (shouldPushToStack || stack.length === 0) {
      stack.push(block)
    }
  }

  const parseProps = (i, block) => {
    let endOfBlockIndex = i
    while (
      endOfBlockIndex < lines.length - 1 &&
      !isBlock(lines[endOfBlockIndex])
    ) {
      endOfBlockIndex++
    }

    const properties = []
    const nested = []

    for (let j = i; j <= endOfBlockIndex; j++) {
      const l = lines[j]
      const line = l.trim()
      const lineIndex = j + 1

      if (isSection(line)) {
        const prop = getSection(line)

        if (isItem(prop) && nested.length > 0) {
          const item = {
            type: 'ArrayItem',
            value: {
              type: 'ObjectExpression',
              properties: [],
              loc: getLoc(lineIndex + 1, 0),
            },
          }

          const last = nested[nested.length - 1]
          if (last.value.type === 'ObjectExpression') {
            last.value = {
              type: 'ArrayExpression',
              elements: [],
              item: prop,
              loc: last.value.loc,
            }
          }

          last.value.elements.push(item)
        } else {
          nested.push({
            type: 'Property',
            loc: getLoc(lineIndex, l.indexOf(prop), l.length - 1),
            key: {
              type: 'Literal',
              value: prop,
              loc: getLoc(
                lineIndex,
                l.indexOf(prop),
                l.indexOf(prop) + prop.length - 1
              ),
            },
            value: {
              type: 'ObjectExpression',
              properties: [],
              loc: getLoc(lineIndex + 1, 0),
            },
          })
        }
      } else if (isProp(line)) {
        const [propRaw, value] = getProp(line)
        const [prop, stemmedTag] = stemStylesFromProp(propRaw)
        const tags = getTags(prop, value)
        if (stemmedTag) {
          tags.push(stemmedTag)
        }

        let last = properties
        if (nested[nested.length - 1]) {
          const lastValue = nested[nested.length - 1].value

          if (lastValue.type === 'ObjectExpression') {
            last = lastValue.properties
          } else if (lastValue.type === 'ArrayExpression') {
            const lastElement =
              lastValue.elements[lastValue.elements.length - 1]
            if (
              lastElement &&
              lastElement.type === 'ArrayItem' &&
              !isItem(prop)
            ) {
              last = lastElement.value.properties
            } else {
              last = lastValue.elements
            }
          }
        }

        if (isItem(prop)) {
          // TODO FIXME cast last to array
          last.push({
            type: 'ArrayItem',
            value: {
              type: 'Literal',
              value: getValue(value),
              loc: getLoc(
                lineIndex,
                l.indexOf(value),
                l.indexOf(value) + value.length - 1
              ),
            },
          })
        } else {
          const propValue = isEmptyList(value)
            ? {
                type: 'ArrayExpression',
                elements: [],
              }
            : {
                type: 'Literal',
                value: getValue(value),
              }

          propValue.loc = getLoc(
            lineIndex,
            l.indexOf(value),
            l.indexOf(value) + value.length - 1
          )

          last.push({
            type: 'Property',
            loc: getLoc(lineIndex, l.indexOf(propRaw), l.length - 1),
            key: {
              type: 'Literal',
              // TODO should we use propRaw as value here?
              value: prop,
              loc: getLoc(
                lineIndex,
                l.indexOf(propRaw),
                l.indexOf(propRaw) + propRaw.length - 1
              ),
            },
            tags,
            meta: getMeta(value, l, lineIndex),
            value: propValue,
          })
        }
      } else if (isEnd(line)) {
        const prevLine = lines[j - 1]

        // TODO close item
        if (isEnd(prevLine)) {
          if (nested.length > 0) {
            const props = nested.pop()

            let last = properties
            if (nested.length > 0) {
              last = nested[nested.length - 1].value.properties
            }
            // TODO add warning
            if (last) {
              last.push(props)
            }
          }
        }
      }
    }

    const loc = getLoc(i + 1, 0, lines[endOfBlockIndex].length - 1)
    loc.end.line = endOfBlockIndex

    while (nested.length > 0) {
      const props = nested.pop()

      let last = properties
      if (nested.length > 0) {
        last = nested[nested.length - 1].value.properties
      }
      if (last) {
        last.push(props)
      }
    }

    if (properties.length > 0) {
      block.properties = {
        type: 'Properties',
        list: properties,
        loc,
      }
    }
  }

  lines.forEach((l, i) => {
    const line = l.trim()
    const lineIndex = i + 1

    if (isBlock(line)) {
      parseBlock(l, i, line, lineIndex)
    } else if (isProp(line) || isSection(line)) {
      let block = stack[stack.length - 1] || views[views.length - 1]
      // TODO add warning
      if (!block) return
      if (block.blocks && block.blocks.list.length > 0) {
        block = block.blocks.list[block.blocks.list.length - 1]
      }

      if (!block.properties) {
        parseProps(i, block)
      }
    } else if (isTodo(line)) {
      // eslint-disable-next-line
      const [_, to, message] = getTodo(line)

      const todo = {
        type: 'Todo',
        loc: getLoc(lineIndex, l.indexOf('#') + 1, l.length - 1),
        to,
        message: message.trim(),
      }

      todos.push(todo)
    } else if (isEnd(line) && stack.length > 0) {
      const prevLine = lines[i - 1]
      if (isEnd(prevLine)) {
        end(stack.pop(), lineIndex)
      }
    }
  })

  if (stack.length > 0) {
    while (!end(stack.pop(), lines.length)) {}
  }

  return {
    fonts,
    todos,
    views,
  }
}
