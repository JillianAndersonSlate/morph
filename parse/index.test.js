import parse from './index.js'

test('#parse', () => {
  VIEWS.forEach(view => expect(parse(view)).toMatchSnapshot())

  expect(parse(WARNING)).toMatchSnapshot()
})

const VIEWS = [
  `Main Vertical`,

  `Main Vertical
Text`,

  `BlueButton Vertical
Before Text

Nested Vertical
ImageInside Image
source https://image.com/file.jpg

VerticalInside Vertical


After Text
text after
color purple
when hover
color red

Last Vertical
backgroundColor blue`,

  // ensure that empty spaces at the end of the block don't block the block from
  // being recognised as a stop point for the props of the previous block
  `A Vertical 
A1 Text
color red

A2 Text    
color white`,

  `Vertical
backgroundColor props.backgroundColor
onClick event => props.onClick(props.name, event)
Title Text
color props red
marginLeft props 10
text props.title This is the title

CaptureText
onFocus props

List
from props
Stuff`,
]

const WARNING = `Warning Vertical
EmptyWhen Vertical
color blue
marginTop props
border 1px solid red
when
color red
when props. stuff
backgroundColor purple
color green`
