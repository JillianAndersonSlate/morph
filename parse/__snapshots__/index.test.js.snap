// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#parse 1`] = `
Object {
  "fonts": Array [],
  "props": Array [],
  "views": Array [
    Object {
      "children": Array [],
      "is": "Main",
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 12,
          "line": 0,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [],
}
`;

exports[`#parse 2`] = `
Object {
  "fonts": Array [],
  "props": Array [],
  "views": Array [
    Object {
      "children": Array [
        Object {
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 3,
              "line": 1,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "name": "Text",
          "properties": Array [],
          "scopes": Array [],
          "type": "Block",
        },
      ],
      "is": "Main",
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 3,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [],
}
`;

exports[`#parse 3`] = `
Object {
  "fonts": Array [],
  "props": Array [],
  "views": Array [
    Object {
      "children": Array [
        Object {
          "is": "Before",
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 2,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "name": "Text",
          "properties": Array [],
          "scopes": Array [],
          "type": "Block",
        },
        Object {
          "children": Array [
            Object {
              "is": "ImageInside",
              "isBasic": true,
              "isGroup": false,
              "loc": Object {
                "end": Object {
                  "column": 0,
                  "line": 6,
                },
                "start": Object {
                  "column": 0,
                  "line": 4,
                },
              },
              "name": "Image",
              "properties": Array [
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 32,
                      "line": 5,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 5,
                    },
                  },
                  "name": "source",
                  "tags": Object {},
                  "type": "Property",
                  "value": "https://image.com/file.jpg",
                },
              ],
              "scopes": Array [],
              "type": "Block",
            },
            Object {
              "children": Array [],
              "is": "VerticalInside",
              "isBasic": true,
              "isGroup": true,
              "isList": false,
              "loc": Object {
                "end": Object {
                  "column": 0,
                  "line": 8,
                },
                "start": Object {
                  "column": 0,
                  "line": 7,
                },
              },
              "name": "Vertical",
              "properties": Array [],
              "scopes": Array [],
              "type": "Block",
            },
          ],
          "is": "Nested",
          "isBasic": true,
          "isGroup": true,
          "isList": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 9,
            },
            "start": Object {
              "column": 0,
              "line": 3,
            },
          },
          "name": "Vertical",
          "properties": Array [],
          "scopes": Array [],
          "type": "Block",
        },
        Object {
          "is": "After",
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 15,
            },
            "start": Object {
              "column": 0,
              "line": 10,
            },
          },
          "name": "Text",
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 9,
                  "line": 11,
                },
                "start": Object {
                  "column": 0,
                  "line": 11,
                },
              },
              "name": "text",
              "tags": Object {},
              "type": "Property",
              "value": "after",
            },
            Object {
              "loc": Object {
                "end": Object {
                  "column": 11,
                  "line": 12,
                },
                "start": Object {
                  "column": 0,
                  "line": 12,
                },
              },
              "name": "color",
              "tags": Object {
                "style": true,
              },
              "type": "Property",
              "value": "purple",
            },
          ],
          "scopes": Array [
            Object {
              "isSystem": true,
              "properties": Array [
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 9,
                      "line": 13,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 13,
                    },
                  },
                  "name": "when",
                  "tags": Object {
                    "scope": "hover",
                    "shouldBeProps": true,
                    "validProps": false,
                  },
                  "type": "Property",
                  "value": "hover",
                },
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 8,
                      "line": 14,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 14,
                    },
                  },
                  "name": "color",
                  "tags": Object {
                    "style": true,
                  },
                  "type": "Property",
                  "value": "red",
                },
              ],
              "value": "hover",
            },
          ],
          "type": "Block",
        },
        Object {
          "children": Array [],
          "is": "Last",
          "isBasic": true,
          "isGroup": true,
          "isList": false,
          "loc": Object {
            "end": Object {
              "column": 19,
              "line": 17,
            },
            "start": Object {
              "column": 0,
              "line": 16,
            },
          },
          "name": "Vertical",
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 19,
                  "line": 17,
                },
                "start": Object {
                  "column": 0,
                  "line": 17,
                },
              },
              "name": "backgroundColor",
              "tags": Object {
                "style": true,
              },
              "type": "Property",
              "value": "blue",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
      ],
      "is": "BlueButton",
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 19,
          "line": 17,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [],
}
`;

exports[`#parse 4`] = `
Object {
  "fonts": Array [],
  "props": Array [],
  "views": Array [
    Object {
      "children": Array [
        Object {
          "is": "A1",
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 3,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "name": "Text",
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 8,
                  "line": 2,
                },
                "start": Object {
                  "column": 0,
                  "line": 2,
                },
              },
              "name": "color",
              "tags": Object {
                "style": true,
              },
              "type": "Property",
              "value": "red",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
        Object {
          "is": "A2",
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 10,
              "line": 5,
            },
            "start": Object {
              "column": 0,
              "line": 4,
            },
          },
          "name": "Text",
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 10,
                  "line": 5,
                },
                "start": Object {
                  "column": 0,
                  "line": 5,
                },
              },
              "name": "color",
              "tags": Object {
                "style": true,
              },
              "type": "Property",
              "value": "white",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
      ],
      "is": "A",
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 10,
          "line": 5,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [
    Object {
      "line": "A Vertical ",
      "loc": Object {
        "start": Object {
          "line": 0,
        },
      },
      "type": "You have some spaces before or after this line. Clean them up.",
    },
    Object {
      "line": "A2 Text    ",
      "loc": Object {
        "start": Object {
          "line": 4,
        },
      },
      "type": "You have some spaces before or after this line. Clean them up.",
    },
  ],
}
`;

exports[`#parse 5`] = `
Object {
  "fonts": Array [],
  "props": Array [
    Object {
      "defaultValue": false,
      "name": "backgroundColor",
      "type": "string",
    },
    Object {
      "defaultValue": "red",
      "name": "color",
      "type": "string",
    },
    Object {
      "defaultValue": 10,
      "name": "marginLeft",
      "type": "number",
    },
    Object {
      "defaultValue": "This is the title",
      "name": "text",
      "type": "string",
    },
    Object {
      "defaultValue": false,
      "name": "onFocus",
      "type": "function",
    },
    Object {
      "defaultValue": false,
      "name": "from",
      "type": "array",
    },
  ],
  "views": Array [
    Object {
      "children": Array [
        Object {
          "is": "Title",
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 7,
            },
            "start": Object {
              "column": 0,
              "line": 3,
            },
          },
          "name": "Text",
          "properties": Array [
            Object {
              "defaultValue": "red",
              "loc": Object {
                "end": Object {
                  "column": 14,
                  "line": 4,
                },
                "start": Object {
                  "column": 0,
                  "line": 4,
                },
              },
              "name": "color",
              "tags": Object {
                "props": true,
                "style": true,
                "validProps": true,
              },
              "type": "Property",
              "value": "props.color",
            },
            Object {
              "defaultValue": 10,
              "loc": Object {
                "end": Object {
                  "column": 18,
                  "line": 5,
                },
                "start": Object {
                  "column": 0,
                  "line": 5,
                },
              },
              "name": "marginLeft",
              "tags": Object {
                "props": true,
                "style": true,
                "validProps": true,
              },
              "type": "Property",
              "value": "props.marginLeft",
            },
            Object {
              "defaultValue": "This is the title",
              "loc": Object {
                "end": Object {
                  "column": 33,
                  "line": 6,
                },
                "start": Object {
                  "column": 0,
                  "line": 6,
                },
              },
              "name": "text",
              "tags": Object {
                "props": true,
                "validProps": true,
              },
              "type": "Property",
              "value": "props.title",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
        Object {
          "isBasic": true,
          "isGroup": false,
          "loc": Object {
            "end": Object {
              "column": 0,
              "line": 10,
            },
            "start": Object {
              "column": 0,
              "line": 8,
            },
          },
          "name": "CaptureText",
          "properties": Array [
            Object {
              "defaultValue": false,
              "loc": Object {
                "end": Object {
                  "column": 12,
                  "line": 9,
                },
                "start": Object {
                  "column": 0,
                  "line": 9,
                },
              },
              "name": "onFocus",
              "tags": Object {
                "props": true,
                "shouldBeProps": true,
                "validProps": true,
              },
              "type": "Property",
              "value": "props.onFocus",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "childrenProxyMap": null,
              "isBasic": false,
              "isGroup": true,
              "isList": false,
              "loc": Object {
                "end": Object {
                  "column": 4,
                  "line": 13,
                },
                "start": Object {
                  "column": 0,
                  "line": 13,
                },
              },
              "name": "Stuff",
              "properties": Array [],
              "scopes": Array [],
              "type": "Block",
            },
          ],
          "isBasic": true,
          "isGroup": true,
          "isList": true,
          "loc": Object {
            "end": Object {
              "column": 4,
              "line": 13,
            },
            "start": Object {
              "column": 0,
              "line": 11,
            },
          },
          "name": "List",
          "properties": Array [
            Object {
              "defaultValue": false,
              "loc": Object {
                "end": Object {
                  "column": 9,
                  "line": 12,
                },
                "start": Object {
                  "column": 0,
                  "line": 12,
                },
              },
              "name": "from",
              "tags": Object {
                "props": true,
                "shouldBeProps": true,
                "validProps": true,
              },
              "type": "Property",
              "value": "props.from",
            },
          ],
          "scopes": Array [],
          "type": "Block",
        },
      ],
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 4,
          "line": 13,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [
        Object {
          "defaultValue": false,
          "loc": Object {
            "end": Object {
              "column": 36,
              "line": 1,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "name": "backgroundColor",
          "tags": Object {
            "props": true,
            "style": true,
            "validProps": true,
          },
          "type": "Property",
          "value": "props.backgroundColor",
        },
        Object {
          "loc": Object {
            "end": Object {
              "column": 48,
              "line": 2,
            },
            "start": Object {
              "column": 0,
              "line": 2,
            },
          },
          "name": "onClick",
          "tags": Object {
            "shouldBeProps": true,
            "validProps": false,
          },
          "type": "Property",
          "value": "event => props.onClick(props.name, event)",
        },
      ],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [
    Object {
      "line": "backgroundColor props.backgroundColor",
      "loc": Object {
        "end": Object {
          "column": 36,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 1,
        },
      },
      "type": "Add a default value to backgroundColor like: backgroundColor props.backgroundColor default value",
    },
    Object {
      "line": "onClick event => props.onClick(props.name, event)",
      "loc": Object {
        "end": Object {
          "column": 48,
          "line": 2,
        },
        "start": Object {
          "column": 0,
          "line": 2,
        },
      },
      "type": "The code you used in the props value is invalid",
    },
    Object {
      "line": "onFocus props",
      "loc": Object {
        "end": Object {
          "column": 12,
          "line": 9,
        },
        "start": Object {
          "column": 0,
          "line": 9,
        },
      },
      "type": "Add a default value to onFocus like: onFocus props default value",
    },
    Object {
      "line": "from props",
      "loc": Object {
        "end": Object {
          "column": 9,
          "line": 12,
        },
        "start": Object {
          "column": 0,
          "line": 12,
        },
      },
      "type": "Add a default value to from like: from props default value",
    },
  ],
}
`;

exports[`#parse 6`] = `
Object {
  "fonts": Array [],
  "props": Array [
    Object {
      "defaultValue": false,
      "name": "marginTop",
      "type": "number",
    },
  ],
  "views": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "is": "EmptyWhen",
          "isBasic": true,
          "isGroup": true,
          "isList": false,
          "loc": Object {
            "end": Object {
              "column": 10,
              "line": 9,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "name": "Vertical",
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 9,
                  "line": 2,
                },
                "start": Object {
                  "column": 0,
                  "line": 2,
                },
              },
              "name": "color",
              "tags": Object {
                "style": true,
              },
              "type": "Property",
              "value": "blue",
            },
            Object {
              "defaultValue": false,
              "loc": Object {
                "end": Object {
                  "column": 14,
                  "line": 3,
                },
                "start": Object {
                  "column": 0,
                  "line": 3,
                },
              },
              "name": "marginTop",
              "tags": Object {
                "props": true,
                "style": true,
                "validProps": true,
              },
              "type": "Property",
              "value": "props.marginTop",
            },
            Object {
              "loc": Object {
                "end": Object {
                  "column": 19,
                  "line": 4,
                },
                "start": Object {
                  "column": 0,
                  "line": 4,
                },
              },
              "name": "border",
              "tags": Object {
                "shorthand": true,
              },
              "type": "Property",
              "value": "1px solid red",
            },
            Object {
              "loc": Object {
                "end": Object {
                  "column": 3,
                  "line": 5,
                },
                "start": Object {
                  "column": 0,
                  "line": 5,
                },
              },
              "name": "when",
              "tags": Object {
                "scope": "",
                "shouldBeProps": true,
                "validProps": false,
              },
              "type": "Property",
              "value": "",
            },
            Object {
              "loc": Object {
                "end": Object {
                  "column": 8,
                  "line": 6,
                },
                "start": Object {
                  "column": 0,
                  "line": 6,
                },
              },
              "name": "color",
              "tags": Object {
                "style": true,
              },
              "type": "Property",
              "value": "red",
            },
          ],
          "scopes": Array [
            Object {
              "isSystem": false,
              "properties": Array [],
              "value": "",
            },
            Object {
              "isSystem": false,
              "properties": Array [
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 16,
                      "line": 7,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 7,
                    },
                  },
                  "name": "when",
                  "tags": Object {
                    "scope": "props. stuff",
                    "shouldBeProps": true,
                    "validProps": false,
                  },
                  "type": "Property",
                  "value": "props. stuff",
                },
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 21,
                      "line": 8,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 8,
                    },
                  },
                  "name": "backgroundColor",
                  "tags": Object {
                    "style": true,
                  },
                  "type": "Property",
                  "value": "purple",
                },
                Object {
                  "loc": Object {
                    "end": Object {
                      "column": 10,
                      "line": 9,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 9,
                    },
                  },
                  "name": "color",
                  "tags": Object {
                    "style": true,
                  },
                  "type": "Property",
                  "value": "green",
                },
              ],
              "value": "props. stuff",
            },
          ],
          "type": "Block",
        },
      ],
      "is": "Warning",
      "isBasic": true,
      "isGroup": true,
      "isList": false,
      "loc": Object {
        "end": Object {
          "column": 10,
          "line": 9,
        },
        "start": Object {
          "column": 0,
          "line": 0,
        },
      },
      "name": "Vertical",
      "properties": Array [],
      "scopes": Array [],
      "type": "Block",
    },
  ],
  "warnings": Array [
    Object {
      "line": "marginTop props",
      "loc": Object {
        "end": Object {
          "column": 14,
          "line": 3,
        },
        "start": Object {
          "column": 0,
          "line": 3,
        },
      },
      "type": "Add a default value to marginTop like: marginTop props default value",
    },
    Object {
      "line": "border 1px solid red",
      "loc": Object {
        "end": Object {
          "column": 19,
          "line": 4,
        },
        "start": Object {
          "column": 0,
          "line": 4,
        },
      },
      "type": "The shorthand border isn't allowed. You need to expand it like:
borderColor
borderStyle
borderWidth",
    },
    Object {
      "line": "when",
      "loc": Object {
        "end": Object {
          "column": 3,
          "line": 5,
        },
        "start": Object {
          "column": 0,
          "line": 5,
        },
      },
      "type": "This when has no condition assigned to it. Add one like: when props.isCondition",
    },
    Object {
      "line": "when props. stuff",
      "loc": Object {
        "end": Object {
          "column": 16,
          "line": 7,
        },
        "start": Object {
          "column": 0,
          "line": 7,
        },
      },
      "type": "The code you used in the props value is invalid",
    },
    Object {
      "line": "backgroundColor purple",
      "loc": Object {
        "end": Object {
          "column": 21,
          "line": 8,
        },
        "start": Object {
          "column": 0,
          "line": 8,
        },
      },
      "type": "You're missing a base prop for backgroundColor. Add it before all whens on the block.",
    },
  ],
}
`;
